### PR TITLE
1624695: Act on changes in upload enable state outside of application

### PR DESCRIPTION
**Pull request by mdboom**
I found a much simpler way forward here that really only adds some slightly different logic in `Glean::new` and a bit of a refactor.  I'm much happier with this than the earlier two approaches.

TODO:
- [x] Port new unit tests to Swift (I don't currently have access to a Mac -- volunteers?)

### CODE REVIEW

File: glean-core/src/lib.rs [hunk 1]
```diff
@@ -308,22 +334,27 @@ impl Glean {
...
     }
 
-    /// Handles the changing of state when upload_enabled changes.
+    /// Handles the changing of state from upload disabled to enabled.
     ///
     /// Should only be called when the state actually changes.
-    /// When disabling, all pending metrics, events and queued pings are cleared.
     ///
-    /// When enabling, the core Glean metrics are recreated.
+    /// The core Glean metrics are recreated.
```
Comment (badboy) [inline glean-core/src/lib.rs]: the docs should call out that the flag is not yet set to `true` when this message should be called.

File: glean-core/src/lib.rs [hunk 2]
```diff
@@ -308,22 +334,27 @@ impl Glean {
...
-    /// # Arguments
+    /// Should only be called when the state actually changes.
     ///
-    /// * `flag` - When true, enable metric collection.
-    fn on_change_upload_enabled(&mut self, flag: bool) {
-        if flag {
-            self.initialize_core_metrics();
-        } else {
-            self.clear_metrics();
+    /// All pending metrics, events and queued pings are cleared.
```
Comment (badboy) [inline glean-core/src/lib.rs]: The docs should call out that it requires the flag to still be true for a deletion request ping to be sent (IMO that could also be a comment on the `if let Err` line below, given that we might not want to document the exact details of the function in its  documentation)

File: glean-core/src/lib.rs (hunk 1 above)
Comment (mdboom) [reply glean-core/src/lib.rs]: But that's not strictly true -- it might be set to true already if this is called during initialization.

Comment (badboy) [reply glean-core/src/lib.rs]: Ah right.
But it is strictly true that afterwards it's definitely set to true.

Comment (mdboom) [reply glean-core/src/lib.rs]: Got it.  Adding this to the comment (and `on_upload_disabled`, too)

**Comment by mdboom** [comment]
This includes a bug fix for the metrics ping scheduler sending on the last day of the month.  Does anyone feel strongly about separating that out?

**Comment by badboy** [comment]
> This includes a bug fix for the metrics ping scheduler sending on the last day of the month. Does anyone feel strongly about separating that out?

As long as it is its own commit, I'm fine.

**Pull request merged**